miner/Miner.hs

```diff
@@ -0,0 +1,285 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE ViewPatterns #-}
+
+-- |
+-- Module: Main
+-- Copyright: Copyright © 2019 Kadena LLC.
+-- License: MIT
+-- Maintainer: Colin Woodbury <colin@kadena.io>
+-- Stability: experimental
+--
+-- The fast "inner loop" of the mining process. Sits idle upon startup until
+-- work is submitted.
+--
+
+module Main ( main ) where
+
+import BasePrelude hiding (app, option)
+import Control.Concurrent.Async (async, race, wait)
+import Control.Concurrent.STM.TMVar
+import Control.Concurrent.STM.TVar (modifyTVar')
+import Control.Error.Util (note)
+import Control.Scheduler (Comp(..), replicateWork, terminateWith, withScheduler)
+import Crypto.Hash.Algorithms (SHA512t_256)
+import Crypto.Hash.IO
+import qualified Data.ByteArray as BA
+import Data.Bytes.Put
+import qualified Data.ByteString as B
+import Data.HashMap.Strict (HashMap)
+import qualified Data.HashMap.Strict as HM
+import qualified Data.Text as T
+import Data.Tuple.Strict (T2(..))
+import Foreign.Marshal.Alloc (allocaBytes)
+import qualified Network.Wai.Handler.Warp as W
+import Options.Applicative
+import Servant.API
+import Servant.Server (Application, Server, serve)
+import Servant.Swagger (toSwagger)
+import Servant.Swagger.UI (SwaggerSchemaUI, swaggerSchemaUIServer)
+
+-- internal modules
+import Chainweb.BlockHeader
+import Chainweb.Difficulty (HashTarget, encodeHashTarget)
+import Chainweb.RestAPI.Orphans ()
+import Chainweb.Time (Micros, Time, encodeTimeToWord64, getCurrentTimeIntegral)
+import Chainweb.Utils (int, runGet)
+import Chainweb.Version (ChainId, ChainwebVersion(..), chainwebVersionFromText)
+import Miner.Types ()
+
+---
+
+--------------------------------------------------------------------------------
+-- Servant
+
+type MinerAPI = "submit" :> ReqBody '[JSON] BlockHeader :> Post '[JSON] ()
+    :<|> "poll" :> Capture "chainid" ChainId
+                :> Capture "blockheight" BlockHeight
+                :> Get '[JSON] (Maybe BlockHeader)
+
+type API = MinerAPI :<|> SwaggerSchemaUI "help" "swagger.json"
```

Comment:
Let's use the common Chainweb UI prefix, which is available via `Chainweb.RestAPI.Utils`.

When we start having different components, we should make sure that are consistent and that the consistency is maintainable by using common utils.